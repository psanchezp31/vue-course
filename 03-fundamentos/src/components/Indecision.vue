<template>
  <h1>Indecision</h1>
  <img v-if="image" :src="image" alt="bg" />
  <div class="bg-dark"></div>
  <div class="indecision-container">
    <input
      v-model="question"
      @keypress="saveInput"
      type="text"
      placeholder="Ask me a question"
    />
    <p>Remember to ends it with a question mark (?)</p>

    <div v-if="isValidQuestion">
      <h2>
        {{ question }}
      </h2>
      <h1>{{ answer }}</h1>
    </div>
  </div>
</template>

<script>
export default {
  data() {
    return {
      question: null,
      answer: null,
      image: null,
      isValidQuestion: false,
    };
  },
  watch: {
    question(value, oldValue) {
      this.isValidQuestion = false;

      console.log({ value });

      if (!value.includes("?")) return;
      this.isValidQuestion = true;
      this.getAnswer();
    },
  },
  methods: {
    async getAnswer() {
      try {
        this.answer = "Pensando...";

        const { answer, image } = await fetch(
          "https://yesno.wtf/api"
        ).then((response) => response.json());

        console.log(answer, image);

        this.answer = answer === "yes" ? "Si!" : "No!";

        this.image = image;

      } catch (error) {
        console.log("IndecisionComponent:", error);
        this.answer = "No se pudo cargar del API";
        this.image = null;
      }
    },
  },
};
</script>

<style scoped>
/*SCOPED hace que el css solo sea aplicado a este componente */
img,
.bg-dark {
  height: 100vh;
  left: 0px;
  max-height: 100%;
  max-width: 100%;
  position: fixed;
  top: 0px;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
  position: relative;
  z-index: 99;
}

input {
  width: 250px;
  padding: 10px 15px;
  border-radius: 5px;
  border: none;
}
input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 20px;
  margin-top: 0px;
}

h1,
h2 {
  color: white;
}

h2 {
  margin-top: 150px;
}
</style>
