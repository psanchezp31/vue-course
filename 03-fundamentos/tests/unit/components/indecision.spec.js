import { shallowMount, mount } from "@vue/test-utils";
import Indecision from "@/components/Indecision";

describe("Indecision Component", () => {
  let wrapper;
  let clgSpy;

  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          answer: "yes",
          forced: false,
          image: "https://yesno.wtf/assets/yes/2.gif",
        }),
    })
  ); //global.fetch.. Global es como el window de js en NODE, ya que node no tiene fetch
  //promise.resolve, se usa para simular el THEN que se le debe dar al fetch. Primero hace el fetch, luego necesita un then.
  //el json() es un método, se mockea con la respuesta que da el promise.resolve

  beforeEach(() => {
    wrapper = shallowMount(Indecision);

    clgSpy = jest.spyOn(console, "log"); //espiar el console.log

    jest.clearAllMocks();
  });

  test("Debe hacer match con el snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("Escribir en el input no debe disparar nada excepto el (console.log)", async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer"); //vm es la instancia de vue. No se puede usar solamente wrapper, porque no es una función.

    const input = wrapper.find("input");
    await input.setValue("hola mundo");

    expect(clgSpy).toHaveBeenCalled();
    expect(getAnswerSpy).toHaveBeenCalledTimes(0);
  });

  test('Escribir el símbolo de interrogación "?" debe disparar el getAnswer', async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer"); //vm es la instancia de vue. No se puede usar solamente wrapper, porque no es una función.

    const input = wrapper.find("input");
    await input.setValue("hola mundo?");

    expect(clgSpy).toHaveBeenCalled();
    expect(getAnswerSpy).toHaveBeenCalledTimes(1);
  });

  test("pruebas en getAnswer", async () => {
    await wrapper.vm.getAnswer();

    const img = wrapper.find("img");

    expect(img.exists()).toBeTruthy();
    expect(wrapper.vm.image).toBe("https://yesno.wtf/assets/yes/2.gif");
    expect(wrapper.vm.answer).toBe("Si!");
  });

  test("pruebas en getAnswer - Fallo en el API", async () => {
    //mock de un catch

    fetch.mockImplementationOnce(() => Promise.reject("API is down"));

    await wrapper.vm.getAnswer();

    const img = wrapper.find("img");

    expect(img.exists()).toBeFalsy();
    expect(wrapper.vm.answer).toBe("No se pudo cargar del API");
  });
});
