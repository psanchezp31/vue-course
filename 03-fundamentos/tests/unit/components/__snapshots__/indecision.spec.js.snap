// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component Debe hacer match con el snapshot 1`] = `
<h1>Indecision</h1>
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Ask me a question">
  <p>Remember to ends it with a question mark (?)</p>
  <!--v-if-->
</div>
`;
